src/lib/strategy-card/strategy-card.js

```diff
@@ -0,0 +1,80 @@
+/*
+ *    Generic Scripting for Strategy Cards.
+ *
+ */
+
+const { globalEvents, Rotator, UIElement } = require("../../wrapper/api");
+const { StrategyCardBorder } = require("./StrategyCardBorder");
+const { PlayerDesk } = require("../../lib/player-desk");
+const locale = require("../../lib/locale");
+let openSelections = {};
+
+function broadcastMessage(message, player) {
+    for (const p of world.getAllPlayers()) {
+        p.sendChatMessage(message, player && player.getPlayerColor());
+    }
+}
+
+function onUiClosedClicked(button, player) {
+    const border = getTopLevelWidget(button);
+
+    // only react on the correct player
+    if (border.getPlayer() !== player) {
+        return;
+    }
+
+    // trigger event for the card itself
+    globalEvents.TI4.onStrategyCardSelectionDone.trigger(border, player);
+
+    // clear internal data and send notifications
+    let selections = openSelections[border.getCard().getId()];
+
+    world.removeUIElement(border.getUI());
+
+    selections.splice(selections.indexOf(border.getPlayer().getSlot()), 1);
+    if (selections.length === 0) {
+        delete openSelections[border.getCard().getId()];
+        broadcastMessage(locale("strategy_card.message.all_resolved"));
+    }
+}
+
+function getTopLevelWidget(element) {
+    const parent = element.getParent();
+    return parent ? getTopLevelWidget(parent) : element;
+}
+
+function createStrategyCardUi(card, widgetFactory) {
+    const playerDesks = PlayerDesk.getPlayerDesks();
+
+    for (const player of world.getAllPlayers()) {
```

Comment:
Making players rather than desks the outer loop means the UI only appears for players currently in the game.  It might be better to iterate over playerDesks and show the UI on the desk even if a player isn't sitting there?  (E.g. they temporarily disconnected?).  I guess you're assuming that the clicking player is the one who owns that UI, which is probably the correct assumption if/when we get on-screen UI.

Reply:
also done

Reply:
I changed the assusmption to check if the player clicking the UI is the player seated at the corresponding desk.